### PR TITLE
Issue 7.chia.1

### PR DESCRIPTION
**Pull request by mahchiahui**
# Pull request questions

## Which issue does this address

Issue number: #7 

## Why was change needed

Add tags and resource group to logically identify resources

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -62,24 +71,36 @@ GOLANG_VERSION ?= 1.13.8
...
+# Terraform Variables
 # -----------------------------------------------------------------------------
 
 TERRAFORM_VERSION ?= 0.12.20
 
 .EXPORT_ALL_VARIABLES:
 TF_VAR_access_key := $(AWS_ACCESS_KEY_ID)
 TF_VAR_secret_key := $(AWS_SECRET_ACCESS_KEY)
-TF_VAR_bucket := $(GIT_ACCOUNT_NAME)-$(GIT_REPOSITORY_NAME)
+TF_VAR_bucket := $(shell echo $(GIT_ACCOUNT_NAME) | tr A-Z a-z)-$(GIT_REPOSITORY_NAME)
```
Comment (bryan-nice) [inline Makefile:94]: Why was this changed? It is necessary for the TF_VAR_query?

File: Makefile [hunk 2]
```diff
@@ -62,24 +71,36 @@ GOLANG_VERSION ?= 1.13.8
...
 TF_VAR_region := $(AWS_DEFAULT_REGION)
 TF_VAR_key := backend/$(GIT_REPOSITORY_NAME).tfstate #Terraform Statefile Name
+TF_VAR_query := {"ResourceTypeFilters": ["AWS::AllSupported"],"TagFilters": [{"Key": "infrastructure-component","Values": ["$(GIT_REPOSITORY_NAME)"]}]}
```
Comment (bryan-nice) [inline Makefile:97]: If you need to transform the GIT_ACCOUNT_NAME variable for TF_VAR_query, then I recommend it being applied here.

File: Makefile [hunk 3]
```diff
@@ -62,24 +71,36 @@ GOLANG_VERSION ?= 1.13.8
 GOLANG_SHA ?= 0567734d558aef19112f2b2873caa0c600f1b4a5827930eb5a7f35235219e9d8
 
 # -----------------------------------------------------------------------------
-# Terraform Varibles
+# Terraform Tag Variables
+# -----------------------------------------------------------------------------
+
+.EXPORT_ALL_VARIABLES:
+TF_VAR_department ?= engineering
```
Comment (bryan-nice) [inline Makefile:78]: This block of TF_VAR's should be combined into one .EXPORT_ALL_VARIABLE block. Please combine them.

File: Makefile [hunk 4]
```diff
@@ -62,24 +71,36 @@ GOLANG_VERSION ?= 1.13.8
...
+TF_VAR_cost_centre ?= 1000
+TF_VAR_infrastructure_component := ${GIT_REPOSITORY_NAME}
```
Comment (bryan-nice) [inline Makefile:80]: In the make file use $(VAR NAME) instead of ${VAR NAME}. The braces are a BASH or Shell construct where as the parentheses are a MAKE construct. Be consistent to avoid collisions.

File: Makefile [hunk 5]
```diff
@@ -62,24 +71,36 @@ GOLANG_VERSION ?= 1.13.8
...
+TF_VAR_infrastructure_version := ${GIT_TAG}
```
Comment (bryan-nice) [inline Makefile:81]: In the make file use $(VAR NAME) instead of ${VAR NAME}. The braces are a BASH or Shell construct where as the parentheses are a MAKE construct. Be consistent to avoid collisions.

File: Makefile [hunk 6]
```diff
@@ -148,6 +169,8 @@ cli:
 		--env AWS_ACCESS_KEY_ID=$(AWS_ACCESS_KEY_ID) \
 		--env AWS_SECRET_ACCESS_KEY=$(AWS_SECRET_ACCESS_KEY) \
 		--env AWS_DEFAULT_REGION=$(AWS_DEFAULT_REGION) \
+		--env TF_VAR_technical_owner=${TF_VAR_technical_owner} \
```
Comment (bryan-nice) [inline Makefile:172]: In the make file use $(VAR NAME) instead of ${VAR NAME}. The braces are a BASH or Shell construct where as the parentheses are a MAKE construct. Be consistent to avoid collisions. Keep in mind this command is being executed in a make target.

File: Makefile [hunk 7]
```diff
@@ -148,6 +169,8 @@ cli:
...
+		--env TF_VAR_business_owner=${TF_VAR_business_owner} \
```
Comment (bryan-nice) [inline Makefile:173]: In the make file use $(VAR NAME) instead of ${VAR NAME}. The braces are a BASH or Shell construct where as the parentheses are a MAKE construct. Be consistent to avoid collisions. Keep in mind this command is being executed in a make target.

File: main.tf [hunk 8]
```diff
@@ -8,19 +8,47 @@ provider "aws" {
...
+    business-owner           = var.business_owner
+    department               = var.department
+    cost-centre              = var.cost_centre
+    infrastructure-component = var.infrastructure_component
+    infrastructure-version   = var.infrastructure_version
+    environment-code         = var.environment_code
+    encryption               = var.encryption
+  }
+}
+
```
Comment (bryan-nice) [inline main.tf:23]: here is a new design pattern to follow.

module "backend-1-1-0" {
  source = "github.com/Senzing/terraform-aws-modules?ref=1.1.0"
}

then following references will be using the same copy.

File: main.tf [hunk 9]
```diff
@@ -8,19 +8,47 @@ provider "aws" {
...
 module "backend_s3_bucket" {
-  source        = "github.com/Senzing/terraform-aws-modules//s3_bucket?ref=1.0.0"
+  source        = "github.com/Senzing/terraform-aws-modules//s3_bucket?ref=1.1.0"
```
Comment (bryan-nice) [inline main.tf:25]: update to leverage the new design pattern mentioned above.

source   = "./.terraform/modules/backend-1-0-0/s3_bucket"

File: main.tf [hunk 10]
```diff
@@ -8,19 +8,47 @@ provider "aws" {
...
   sse_algorithm = "AES256"
   force_destroy = true
+
+  # tags
+  tags = data.null_data_source.common_tags.outputs
 }
 
 module "backend_dynamodb" {
-  source         = "github.com/Senzing/terraform-aws-modules//dynamodb_table?ref=1.0.0"
+  source         = "github.com/Senzing/terraform-aws-modules//dynamodb_table?ref=1.1.0"
```
Comment (bryan-nice) [inline main.tf:36]: update to leverage the new design pattern mentioned above.

source = "./.terraform/modules/backend-1-0-0/dynamodb_table"

File: main.tf [hunk 11]
```diff
@@ -8,19 +8,47 @@ provider "aws" {
...
   attribute_type = "S"
-}
\ No newline at end of file
+
+  # tags
+  tags = data.null_data_source.common_tags.outputs
+}
+
+module "backend_resource_group" {
+  source = "github.com/Senzing/terraform-aws-modules//resource_groups?ref=1.1.0"
```
Comment (bryan-nice) [inline main.tf:48]: update to leverage the new design pattern mentioned above.

source = "./.terraform/modules/backend-1-0-0/resource_groups"

File: Makefile [hunk 12]
```diff
@@ -30,6 +30,15 @@ endif
 ifndef AWS_DEFAULT_REGION
 $(error $(BOLD)$(RED)AWS_DEFAULT_REGION is not defined, please set this environment variable before proceeding$(RESET))
 endif
+
+ifndef TF_VAR_technical_owner
```
Comment (bryan-nice) [inline Makefile:34]: these should not cause an error and exit from the make process. Default these variables if a user did not set them.

File: Makefile [hunk 13]
```diff
@@ -30,6 +30,15 @@ endif
 ifndef AWS_DEFAULT_REGION
```
Comment (bryan-nice) [inline Makefile:30]: Update this logic to force a user to log in to aws with this command

aws configure

reference to understand more details https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html.

If a user did not provide the necessary credentials, then error and exit the make process.

File: Makefile (hunk 1 above)
Comment (mahchiahui) [reply Makefile:94]: yes, the s3 bucket requires that its name contain only lowercase character. In our case, that would be "Senzing-terraform-aws-backend". The extra code I wrote would change "Senzing" to "senzing". This would result in the S3 bucket name being "senzing-terraform-aws-backend", which would confirm with the S3 bucket naming convention

Comment (bryan-nice) [reply Makefile:94]: i suggest applying the lowercase logic at the source variable level it is consistent threw out the makefile

Comment (mahchiahui) [reply Makefile:94]: got it
